core/src/test/java/io/lionweb/lioncore/java/utils/LanguageValidatorTest.java

```diff
@@ -303,4 +303,71 @@ public void anSDTWithNoFieldsIsInvalid() {
+    sdt.addField(new Field("MyStringField", LionCoreBuiltins.getString(), "fs_id", "fs_key"));
+    assertEquals(Collections.emptySet(), l.validate().getIssues());
+
+    sdt.addField(new Field("MyCircularField", sdt, "fc_id", "fc_key"));
+    Set<Issue> issuesB = l.validate().getIssues();
+    assertEquals(1, issuesB.size());
+    Issue issueB0 = issuesB.iterator().next();
+    assertEquals(
+        new Issue(
+            IssueSeverity.Error, "Circular references are forbidden in StructuralDataFields", sdt),
+        issueB0);
+  }
+
+  @Test
+  public void anSDTShouldNotHaveIndirectlyCircularReferences() {
+    Language l = new Language("MyLanguage", "my_language_id", "my_language_key");
+    StructuredDataType sdtA = new StructuredDataType(l, "SDTA", "sdta_id", "sdta_key");
+    StructuredDataType sdtB = new StructuredDataType(l, "SDTB", "sdtb_id", "sdtb_key");
+    sdtA.addField(new Field("MySDTField", sdtB, "fsdt1_id", "fsdt1_key"));
+    sdtB.addField(new Field("MyStringField", LionCoreBuiltins.getString(), "fs_id", "fs_key"));
+    assertEquals(Collections.emptySet(), l.validate().getIssues());
+
+    sdtB.addField(new Field("MyCircularField", sdtA, "fc_id", "fc_key"));
+    Set<Issue> issuesB = l.validate().getIssues();
+    assertEquals(2, issuesB.size());
+    assert (issuesB.contains(
+        new Issue(
+            IssueSeverity.Error,
+            "Circular references are forbidden in StructuralDataFields",
+            sdtA)));
+    assert (issuesB.contains(
+        new Issue(
+            IssueSeverity.Error,
+            "Circular references are forbidden in StructuralDataFields",
+            sdtB)));
+  }
+
+  @Test
+  public void checkDirectCircularityOfSDTs() {
+    Language l = new Language("MyLanguage", "my_language_id", "my_language_key");
+    StructuredDataType sdt = new StructuredDataType(l, "SDT", "sdt_id", "sdt_key");
+    sdt.addField(new Field("MyStringField", LionCoreBuiltins.getString(), "fs_id", "fs_key"));
+    assertFalse(LanguageValidator.isCircular(sdt));
+
+    sdt.addField(new Field("MyCircularField", sdt, "fc_id", "fc_key"));
+    assertTrue(LanguageValidator.isCircular(sdt));
+  }
+
+  @Test
+  public void checkIndirectCircularityOfSDTs() {
+    Language l = new Language("MyLanguage", "my_language_id", "my_language_key");
+    StructuredDataType sdtA = new StructuredDataType(l, "SDTA", "sdta_id", "sdta_key");
+    StructuredDataType sdtB = new StructuredDataType(l, "SDTB", "sdtb_id", "sdtb_key");
+    sdtA.addField(new Field("MySDTField", sdtB, "fsdt1_id", "fsdt1_key"));
+    sdtB.addField(new Field("MyStringField", LionCoreBuiltins.getString(), "fs_id", "fs_key"));
+    assertFalse(LanguageValidator.isCircular(sdtA));
+    assertFalse(LanguageValidator.isCircular(sdtB));
+
+    sdtB.addField(new Field("MyCircularField", sdtA, "fc_id", "fc_key"));
+    assertTrue(LanguageValidator.isCircular(sdtA));
+    assertTrue(LanguageValidator.isCircular(sdtB));
+  }
```

Comment:
maybe add a test with three-way circle? SDT `a` contains SDT `b`, `b` contains SDT `c`, and `c` contains `a`?

Reply:
Good idea: I added such test